### ISSUE TITLE
stream.hls: Apple Low-Latency HLS Support (ALHLS)

### ISSUE THREAD

**Issue by Ro-Den** [streamlink/streamlink#4054]
### Checklist

- [X] This is a feature request and not a different kind of issue
- [X] [I have read the contribution guidelines](https://github.com/streamlink/streamlink/blob/master/CONTRIBUTING.md#contributing-to-streamlink)
- [X] [I have checked the list of open and recently closed plugin requests](https://github.com/streamlink/streamlink/issues?q=is%3Aissue+label%3A%22feature+request%22)

### Description

[Enabling Low-Latency HLS](https://developer.apple.com/documentation/http_live_streaming/enabling_low-latency_hls)

Is it planned in the near future?

I've tried amending `hls_playlist.py` only but stuck at `media_sequence` which is not used in this case, as I understand.

The only way around is to use `PRELOAD-HINT` but it breaks the whole logic of counting segments.

**Comment by bastimeyer** [comment]
ALHLS hasn't been implemented yet, because it's not part of the official HLS standard from 2017 (RFC 8216 - based on Apple's initial HLS documents). On top of that, lots of sites/services implement their own low latency streaming, like Twitch for example, which uses custom [`EXT-X-TWITCH-PRELOAD`](https://github.com/streamlink/streamlink/blob/2.4.0/src/streamlink/plugins/twitch.py#L34-L37) tags for future prefetch segments, because of lack of proper standardization.

Since 2019 there are two competing low latency HLS protocols:
- LHLS (a collaboration effort between various service/software providers and player implementers)
- ALHLS (Apple's new HLS revision)

https://mux.com/blog/the-community-gave-us-low-latency-live-streaming-then-apple-took-it-away/

Implementing ALHLS means adding support for lots of new tags and logic, like `EXT-X-SERVER-CONTROL` for example for server capability negotiation, the HTTP GET parameters that come with prefetch HLS segments, partial segments which are independent of the media sequence, and lots of other stuff.

LHLS on the other hand uses simple HTTP/1.1 responses with chunked encoding, meaning this is already handled automatically. Streamlink has recently switched to streamed segments by default, so `--hls-segment-stream-data` is not a requirement anymore and segment data will be written to the output buffer before the entire segment has been downloaded.

**Comment by Ro-Den** [comment]
Tags are handled automatically if I'm not mistaken. So, adding them is not a problem.

Apart of `EXT-X-SERVER-CONTROL` and other BS, ALHLS is just 2 chunk lists combined together for whatever reason (I guess, to produce a new standard). The only problem is that the 2nd chunk list uses another logic which doesn't rely on the media sequence.

**Comment by bastimeyer** [comment]
> So, adding them is not a problem.

ALHLS adds new tags with special logic, and all of this needs to be implemented. This doesn't come for free, because the low latency edge segments are split into multiple indexed partial segments (and can be merged by the server later on), which has quite an impact on how the `SegmentedStreamWorker`, `SegmentedStreamWriter`, `HLSStreamWorker` and `HLSStreamWriter` work.

In addition to that, playlist reloads can be blocked by the server to reduce the amount of polling and segments can take optional parameters for signalling a blocking request as well.

Playlist delta updates should be rather trivial.

Then there's also variable substitutions via `EXT-X-DEFINE` and the `{$var}` syntax that come with recent versions of HLS which hasn't been implemented yet. Same with a couple of other stuff (irrelevant to LL streaming though).

----

This is the current draft version of "ALHLS" which will make it into a new standard eventually. The current iteration is version 10 (RFC 8216 had 25 iterations). As long as it's still not officially standardized, it doesn't make too much sense implementing it yet considering that stuff can/will get changed unless the core parts are considered stable.

https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis

> Abstract
> 
>    This document obsoletes RFC 8216.  It describes a protocol for
>    transferring unbounded streams of multimedia data.  It specifies the
>    data format of the files and the actions to be taken by the server
>    (sender) and the clients (receivers) of the streams.  It describes
>    version 10 of this protocol.
> 
> 
> Status of This Memo
> 
>    This Internet-Draft is submitted in full conformance with the
>    provisions of BCP 78 and BCP 79.
> 
>    Internet-Drafts are working documents of the Internet Engineering
>    Task Force (IETF).  Note that other groups may also distribute
>    working documents as Internet-Drafts.  The list of current Internet-
>    Drafts is at https://datatracker.ietf.org/drafts/current/.
> 
>    Internet-Drafts are draft documents valid for a maximum of six months
>    and may be updated, replaced, or obsoleted by other documents at any
>    time.  It is inappropriate to use Internet-Drafts as reference
>    material or to cite them other than as "work in progress."
> 
>    This Internet-Draft will expire on October 29, 2021.

**Comment by Ro-Den** [comment]
> ALHLS adds new tags with special logic

The only really special logic is that it relies on `PRELOAD-HINT` instead of `MEDIA-SEQUENCE`.

Another "special" logic, I guess, is if you have a poor connection you should either try to switch to a lower bit-rate or switch to "legacy" HLS or get blocked by the server.

And by the way, why do we still use the "best/worst" stream logic? Isn't it possible to switch quality/bit-rate on the fly? It is bad for recording, I guess. But we are against copyright infringement, aren't we?

> This Internet-Draft will expire on October 29, 2021.

OK, let's wait and see if it will be officially standardized or not.

**Comment by Ro-Den** [comment]
The draft has expired. Now they would submit one more version and wait another 6 months.

It took Apple more than 8 years and 23 versions to get the 1st Edition of HTTP Live Streaming approved. So, the 2nd Edition may be approved in 4 years.

@beardypig now most players support HLS almost flawlessly. MPV can even mix audio and video tracks from different sub-streams (variants) on the fly. Or you can load m3u8 file as a playlist. The only advantage of Streamlink is the automatic link extraction, which is good for the automation of recording. As for viewing it's much easier to sniff links in a browser.

**Comment by beardypig** [comment]
@Ro-Den, you're right they do support HLS. But Streamlink doesn't output an HLS, it outputs video streams which typically do not change resolution. I'm not sure if there is a container format that supports resolution changes mid file (it would not be a very common use case for regular video file, which is what Streamlink produces).
